### PR TITLE
Various attempts to fix and improve the txn manager

### PR DESCRIPTION
**Pull request by Vagabond**
This is a draft because this is a grab-bag of changes I made while trying to find and fix bugs in the txn manager. Some of the changes don't make sense to merge, but to aid discussion I'm going to open this draft PR. I will attempt to explain the rationale for some of my changes, and flag ones that should probably not merge in comments on the code.

### CODE REVIEW

File: src/transactions/blockchain_txn_mgr.erl [hunk 1]
```diff
@@ -274,25 +298,24 @@ signatory_rand_members(Chain, SubmitF, Acceptions, Rejections) ->
             %% so use a random consensus member
```
Comment (evanmcc) [inline src/transactions/blockchain_txn_mgr.erl:298]: imo this case should just make the group, and move the processing below it, as it looks like somehow we don't shuffle on the signatory side (the common case).

File: src/transactions/blockchain_txn_mgr.erl [hunk 2]
```diff
@@ -422,7 +465,7 @@ delete_cached_txn(Txn) ->
 -spec cached_txn(blockchain_txn:txn())-> {ok, cached_txn_type()} | {error, txn_not_found}.
 cached_txn(Txn)->
     case ets:lookup(?TXN_CACHE, Txn) of
-        [Txn] -> {ok, Txn};
+        [Res] -> {ok, Res};
```
Comment (Vagabond) [inline src/transactions/blockchain_txn_mgr.erl:468]: This was the core bug and it was breaking the removing of dialers. The only way the txn could clear was to appear in a block, a rejection or a failure would never allow the txn to be offered to other members.

File: src/transactions/blockchain_txn_mgr.erl [hunk 3]
```diff
@@ -110,50 +120,64 @@ handle_cast({submit, Txn, Callback}, State=#state{chain = undefined}) ->
...
 
-handle_info({dial_failed, {Dialer, Txn, Member}}, #state{chain = Chain, submit_f = SubmitF} = State) ->
+handle_info({dial_failed, {Dialer, Txn, Member}}, State) ->
     lager:debug("txn: ~s, dial_failed: ~p, Dialer: ~p", [blockchain_txn:print(Txn), Member, Dialer]),
-    ok = retry(Txn, Chain, SubmitF),
+    ok = retry(Txn, Dialer),
+    {noreply, State};
+
+handle_info({timeout, {Dialer, Txn, Member}}, State) ->
+    lager:debug("txn: ~s, timeout: ~p, Dialer: ~p", [blockchain_txn:print(Txn), Member, Dialer]),
```
Comment (evanmcc) [inline src/transactions/blockchain_txn_mgr.erl:163]: Should we kill the dialer here?  we clean up the state in retry, but don't seem to do that.  what happens when we remove the dialer but then get a different message from it (e.g. the timeout and a success both go into the message queue for this process before we manage to process either)?  Are we idempotent both ways?

File: src/transactions/blockchain_txn_mgr.erl [hunk 4]
```diff
@@ -307,34 +330,51 @@ process_cached_txns(Chain, CurBlockHeight, SubmitF, _Sync, IsNewElection, NewGro
     {ValidTransactions, InvalidTransactions} = blockchain_txn:validate(Txns, Chain),
     ok = lists:foreach(
         fun({Txn, {Callback, RecvBlockHeight, Acceptions, Rejections, Dialers}}) ->
-            case {lists:member(Txn, InvalidTransactions), lists:member(Txn, ValidTransactions)} of
-                {false, false} ->
+            IsValidStandalone = case blockchain_txn:is_valid(Txn, Chain) of
```
Comment (Vagabond) [inline src/transactions/blockchain_txn_mgr.erl:333]: This relies on my changes to check nonces in is_valid and probably should not go in. I have a related branch of core that has a function to determine dependencies between transactions which we could probably use here instead.

File: src/transactions/blockchain_txn_mgr.erl [hunk 5]
```diff
@@ -274,25 +298,24 @@ signatory_rand_members(Chain, SubmitF, Acceptions, Rejections) ->
...
-retry(Txn, Chain, SubmitF) ->
+-spec retry(blockchain_txn:txn(), pid()) -> ok.
+retry(Txn, Dialer) ->
     case cached_txn(Txn) of
         {error, _} ->
             %% We no longer have this txn, do nothing
             ok;
-        {ok, {Txn, {Callback, RecvBlockHeight, Acceptions, Rejections, _Dialers}}} ->
-            %% resubmit the txn to CG
-            NewDialers = submit_txn_to_cg(Chain, Txn, SubmitF, Acceptions, Rejections),
```
Comment (Vagabond) [inline src/transactions/blockchain_txn_mgr.erl:294]: I removed all the submit_txn_to_cg calls except for the one called from process_cached_txns to simplifiy things while debugging. Additionally process_cached_txns had a bunch of logic not replicated on the other call paths, so we should consider that before blindly restoring the other submit calls.

File: src/transactions/blockchain_txn_mgr.erl (hunk 2 above)
Comment (andymck) [reply src/transactions/blockchain_txn_mgr.erl:472]: fecking bassdrinker hiding in plain sight

File: src/transactions/blockchain_txn_mgr.erl [hunk 6]
```diff
@@ -189,14 +213,12 @@ code_change(_OldVsn, State, _Extra) ->
...
-    SubmitF = submit_f(length(Signatories)),
-    RejectF = reject_f(length(Signatories)),
+    {ok, N} = blockchain:config(?num_consensus_members, blockchain:ledger(Chain)),
+    SubmitF = submit_f(N),
+    RejectF = reject_f(N),
     %% process any cached txn from before we had a chain, none of these will have been submitted as yet
-    F = fun({Txn, {Callback, _RecvBlockHeight, _Acceptions, _Rejections, _Dialers}}) ->
-            NewDialers = submit_txn_to_cg(Chain, Txn, SubmitF, [], []),
-            ok = cache_txn(Txn, Callback, Height, [], [], NewDialers)
+    F = fun({Txn, {Callback, _RecvBlockHeight, Acceptions, Rejections, Dialers}}) ->
```
Comment (andymck) [inline src/transactions/blockchain_txn_mgr.erl:220]: If we end up not submitting the txns from here in the final changes and only doing so from the add_block event then we can remove this fun and its execution completely

File: src/transactions/blockchain_txn_dialer.erl [hunk 7]
```diff
@@ -50,12 +51,13 @@ dial(Pid) ->
 init(Args) ->
     lager:debug("blockchain_txn_dialer started with ~p", [Args]),
     [Parent, Txn, Member] = Args,
-    {ok, #state{parent=Parent, txn=Txn, member=Member}}.
+    Ref = erlang:send_after(30000, Parent, {timeout, {self(), Txn, Member}}),
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/transactions/blockchain_txn_dialer.erl:54]: This is if the dialer is stuck somehow (libp2p dials might hang?). Sending the timeout directly to the parent means that any weird libp2p failure (or a process crash) cannot interfere with the delivery of this message. The only way this message gets stopped is by the dialer cancelling the timer.

File: src/transactions/blockchain_txn_mgr.erl [hunk 8]
```diff
@@ -206,6 +228,7 @@ initialize_with_chain(State, Chain)->
 handle_add_block_event({add_block, BlockHash, Sync, _Ledger}, State=#state{chain = Chain,
                                                                            cur_block_height = CurBlockHeight})->
     #state{submit_f = SubmitF, chain = Chain} = State,
+    HasBeenSynced = Sync == false orelse State#state.has_been_synced,
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/transactions/blockchain_txn_mgr.erl:231]: if you've managed to sync up at least once, the rationale is you're probably close enough to the head that your transactions are reasonably valid

File: src/transactions/blockchain_txn_mgr.erl (hunk 1 above)
Comment (Vagabond) [reply src/transactions/blockchain_txn_mgr.erl:298]: Is this still relevant?

File: src/transactions/blockchain_txn_mgr.erl (hunk 3 above)
Comment (Vagabond) [reply src/transactions/blockchain_txn_mgr.erl:163]: Yes, we should stop the dialer here. Yes, there's probably a race condition here.

Comment (andymck) [reply src/transactions/blockchain_txn_mgr.erl:163]: i have added the stop as part of my branch to use records

File: src/transactions/blockchain_txn_mgr.erl (hunk 6 above)
Comment (andymck) [reply src/transactions/blockchain_txn_mgr.erl:220]: I commented this fun out for now as it serves no purpose.  We should uncomment if we readd submitting of txns as part of the submit api handling

**Pull request merged**